### PR TITLE
SPI APA102 - multi port fix

### PR DESCRIPTION
**Pull request by s-light**
VIA IRC:
VoltVisionSteve found a bug: 
the APA102 code only worked with one port.
should be fixed now.
protocol APA102_START_FRAME_BYTES are only added for m_output_number==0

### CODE REVIEW

File: plugins/spi/SPIOutput.cpp [hunk 1]
```diff
@@ -522,8 +522,13 @@ void SPIOutput::IndividualAPA102Control(const DmxBuffer &buffer) {
 
   // We always check out the entire string length, even if we only have data
   // for part of it
-  const uint16_t output_length = APA102_START_FRAME_BYTES +
+  // only add the APA102_START_FRAME_BYTES on the first port!!
+  if (m_output_number == 0) {
+    const uint16_t output_length = APA102_START_FRAME_BYTES +
       (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL);
+  } else {
+    const uint16_t output_length = (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL);
```
Comment (nomis52) [inline plugins/spi/SPIOutput.cpp]: I'd change this to:

const uint16_t output_length = (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL);
if (m_output_number ==0) {
  output_length += APA102_START_FRAME_BYTES;
}

I think that makes it clearer what's going on.

File: plugins/spi/SPIOutput.cpp [hunk 2]
```diff
@@ -534,17 +539,26 @@ void SPIOutput::IndividualAPA102Control(const DmxBuffer &buffer) {
...
 
-    // We need to avoid the first 4 bytes of the buffer since that acts as a
-    // start of frame delimiter
-    unsigned int spi_offset = APA102_START_FRAME_BYTES +
+
+    // only skip APA102_START_FRAME_BYTES on the first port!!
+    if (m_output_number == 0) {
+      // We need to avoid the first 4 bytes of the buffer since that acts as a
+      // start of frame delimiter
+      uint16_t spi_offset = APA102_START_FRAME_BYTES +
```
Comment (nomis52) [inline plugins/spi/SPIOutput.cpp]: Same here.

File: plugins/spi/SPIOutput.cpp [hunk 3]
```diff
@@ -579,8 +593,13 @@ void SPIOutput::CombinedAPA102Control(const DmxBuffer &buffer) {
   }
 
   // get data for entire string length
-  const uint16_t output_length = APA102_START_FRAME_BYTES +
+  // only add the APA102_START_FRAME_BYTES on the first port!!
+  if (m_output_number == 0) {
+    const uint16_t output_length = APA102_START_FRAME_BYTES +
       (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL);
+  } else {
+    const uint16_t output_length = (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL);
```
Comment (nomis52) [inline plugins/spi/SPIOutput.cpp]: and here

File: plugins/spi/SPIOutput.cpp (hunk 1 above)
Comment (s-light) [inline plugins/spi/SPIOutput.cpp]: hmm - i think i cant add to a const?!

Comment (nomis52) [inline plugins/spi/SPIOutput.cpp]: Yes sorry, Iif you want to keep the const do:

const uint16_t output_length = (m_pixel_count * APA102_SPI_BYTES_PER_PIXEL) + 
                                                 (m_output_number ? 0 : APA102_START_FRAME_BYTES);

**Comment by peternewman** [comment]
Can we have a new test case for this condition please @s-light . You'll need to fix the Travis errors too.

**Comment by s-light** [comment]
currently its building on my RPI - if this is ready i will test with hardware and report it here.

File: plugins/spi/SPIOutputTest.cpp [hunk 4]
```diff
@@ -681,4 +710,35 @@ void SPIOutputTest::testCombinedAPA102Control() {
...
+  SPIOutput output1(m_uid, &backend, option1);
+  // set personality
+  output1.SetPersonality(this_test_personality);
+  // setup some 'DMX' data
+  buffer.SetFromString("1, 10, 100");
+  // simulate incoming data
+  output1.WriteDMX(buffer);
+  // get fake SPI data stream
+  data = backend.GetData(1, &length);
+  OLA_WARN << "data:";
```
Comment (peternewman) [inline plugins/spi/SPIOutputTest.cpp]: I think you can remove this debugging now.

**Comment by peternewman** [comment]
LGTM

File: plugins/spi/SPIOutputTest.cpp (hunk 4 above)
Comment (s-light) [inline plugins/spi/SPIOutputTest.cpp]: you are right - i have missed this - sorry...

**Comment by s-light** [comment]
its working! :-)

**Comment by nomis52** [comment]
LGTM, thanks for the fix.

**Pull request merged**